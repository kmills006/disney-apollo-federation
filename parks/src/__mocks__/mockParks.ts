export const MOCK_PARKS = [
  {
    name: 'Park One',
    permalink: 'park-one',
  },
  {
    name: 'Park Two',
    permalink: 'park-two',
  },
];
