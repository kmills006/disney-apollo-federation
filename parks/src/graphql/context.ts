import {
  parkRepository,
  IParkRepository,
} from '../repositories/parks';

export interface IResolverContext {
  repositories: {
    parks: IParkRepository;
  };
}

export const context = (): IResolverContext => ({
  repositories: {
    parks: parkRepository(),
  },
});
