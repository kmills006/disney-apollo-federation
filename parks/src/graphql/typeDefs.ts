import { gql } from 'apollo-server';

export const typeDefs = gql`
  type Query {
    _placeholder: String!

    parks: [Park!]
  }

  type Park @key(fields: "permalink") {
    "Walt Disney World Park name"
    name: String!

    "Unchanged URL for the park."
    permalink: String!
  }
`;
