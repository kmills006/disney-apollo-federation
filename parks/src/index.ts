console.log('Hello from the Parks domain!');
