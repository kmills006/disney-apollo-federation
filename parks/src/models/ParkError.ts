export interface IParkError {
  cause: Error;
  message: string;
}

export type IGraphQLError = Pick<IParkError, 'message'>;
