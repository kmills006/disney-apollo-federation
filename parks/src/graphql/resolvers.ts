import { pipe } from 'fp-ts/lib/pipeable';
import * as O from 'fp-ts/lib/Option';
import { GraphQLResolveInfo } from 'graphql/type';

import { IPark, NoParkFoundError } from '@models/Park';
import { IResolverContext } from './context';

type ResolveType<P, T> = (parent: P) => T | null;

type ResolverFunc<R, A = {}, P = {}> = (
  parent: P,
  args: A,
  ctx: IResolverContext,
  info: GraphQLResolveInfo
) => R;

type ParkPayload = IPark | NoParkFoundError;
type ParkPayloadTypeName = 'Park' | 'NoParkFoundError';

interface IQueryResolvers {
  parks: ResolverFunc<IPark[]>;

  park: ResolverFunc<ParkPayload, { permalink: string }>
}

interface IParkPayloadResolvers {
  __resolveType: ResolveType<ParkPayload, ParkPayloadTypeName>;
}

export interface IResolvers {
  Query: IQueryResolvers;
  ParkPayload: IParkPayloadResolvers;
}

export const findTypeNameInParent = <T extends string, P = {}>(
  parent: P,
  mapping: string[][],
): T | null => {
  const [, typename] = mapping.find(([property]) => property in parent) || [null, null];

  return typename as T;
};

export const resolvers: IResolvers = {
  ParkPayload: {
    __resolveType: (parent) => findTypeNameInParent<ParkPayloadTypeName>(parent, [
      ['name', 'Park'],
      ['message', 'NoParkFoundError'],
    ]),
  },

  Query: {
    parks: (_, __, ctx) => ctx.repositories.parks.getParks(),

    park: (_, args, ctx) => pipe(
      ctx.repositories.parks.getParkByPermalink(args.permalink),
      O.getOrElseW(() => ({
        message: `Park with permalink [${args.permalink}] not found.`,
        permalink: args.permalink,
      })),
    ),
  },
};
