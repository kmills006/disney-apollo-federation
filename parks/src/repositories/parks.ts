import { fold, tryCatch, Either } from 'fp-ts/lib/Either';
import { fromNullable, Option } from 'fp-ts/lib/Option';
import { pipe } from 'fp-ts/lib/pipeable';

import { IPark } from '@models/Park';
import { IParkError } from '@models/ParkError';
import { handleParkError } from '../helpers/handleParkError';

// TODO: Fix module path
import { parseJsonFile } from '../db';

type IParkRepositoryError = IParkError;

export interface IParkRepository {
  getParks: () => Either<IParkRepositoryError, IPark[]>;
  getParkByPermalink: (permalink: string) => Option<IPark>;
}

export const repository = (parks: IPark[]) => (): IParkRepository => ({
  getParks: () => tryCatch(
    () => parks,
    (e) => handleParkError(e as Error),
  ),

  getParkByPermalink: (permalink) => (
    fromNullable(parks.find((p) => p.permalink === permalink))
  ),
});

const parks = pipe(
  parseJsonFile<IPark[]>('src/db/parks.json'),
  fold(
    (e) => {
      console.warn(`Unable to parse parks.json: ${e.message}`);
      return [];
    },
    (p) => p,
  ),
);

console.log('parks', parks);

export const parkRepository = repository(parks);
