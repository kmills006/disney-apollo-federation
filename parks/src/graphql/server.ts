import { buildFederatedSchema } from '@apollo/federation';
import { ApolloServer } from 'apollo-server';

import { context } from './context';
import { resolvers } from './resolvers';
import { typeDefs } from './typeDefs';

const apolloServer = () => async () => {
  const schema = buildFederatedSchema({ resolvers, typeDefs });

  const server = new ApolloServer({
    schema,
    context: () => context(),
  });

  server.listen(4000).then(({ url }) => {
    console.log(`ðŸš€ Server ready at ${url}`);
  });
};

export const startApolloServer = apolloServer();
