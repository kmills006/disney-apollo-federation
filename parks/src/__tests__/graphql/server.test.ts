import { ApolloServer, gql } from 'apollo-server';
import { createTestClient } from 'apollo-server-testing';
import * as E from 'fp-ts/lib/Either';
import * as O from 'fp-ts/lib/Option';
import { mock, mockDeep } from 'jest-mock-extended';

import { MOCK_PARKS } from '../../__mocks__/mockParks';
import { resolvers } from '../../graphql/resolvers';
import { typeDefs } from '../../graphql/typeDefs';
import { constructApolloServer } from '../../graphql/server';
import { IParkRepository } from '../../repositories/parks';
import { IParkError } from '../../models/ParkError';

describe('graphql', () => {
  describe('server', () => {
    const mockParksRepository = mockDeep<IParkRepository>();
    const mockError = mock<IParkError>({ message: 'forced error' });

    let server: ApolloServer;

    beforeEach(() => {
      server = constructApolloServer({
        resolvers,
        typeDefs,
        context: () => ({
          repositories: { parks: mockParksRepository },
        }),
      });
    });

    describe('queries', () => {
      describe('parks', () => {
        const GET_PARKS = gql`query { parks { name permalink } }`;

        it('returns a list of parks', async () => {
          mockParksRepository.getParks.mockReturnValueOnce(E.right(MOCK_PARKS));

          const { query } = createTestClient(server);
          const actual = await query({ query: GET_PARKS });

          expect(actual).toMatchSnapshot();
        });

        it('returns an ApolloError', async () => {
          mockParksRepository.getParks.mockReturnValueOnce(E.left(mockError));

          const { query } = createTestClient(server);
          const actual = await query({ query: GET_PARKS });

          expect(actual).toMatchSnapshot();
        });
      });

      describe('park', () => {
        const GET_PARK = gql`
          query Park($permalink: String!) {
            park(permalink: $permalink) {
              ... on Park { name permalink }
              ... on NoParkFoundError { message permalink }
            }
          }
        `;

        const mockPark = MOCK_PARKS[0];

        it('returns a park by its permalink', async () => {
          mockParksRepository.getParkByPermalink.mockReturnValueOnce(O.some(mockPark));

          const { query } = createTestClient(server);
          const actual = await query({
            query: GET_PARK,
            variables: { permalink: mockPark.permalink },
          });

          expect(actual).toMatchSnapshot();
        });

        it('returns a NoParkFoundError', async () => {
          mockParksRepository.getParkByPermalink.mockReturnValueOnce(O.none);

          const { query } = createTestClient(server);
          const actual = await query({
            query: GET_PARK,
            variables: { permalink: mockPark.permalink },
          });

          expect(actual).toMatchSnapshot();
        });
      });
    });
  });
});
