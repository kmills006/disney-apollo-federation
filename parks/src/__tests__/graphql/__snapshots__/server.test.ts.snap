// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`graphql server queries park returns a NoParkFoundError 1`] = `
Object {
  "data": Object {
    "park": Object {
      "message": "Park with permalink [park-one] not found.",
      "permalink": "park-one",
    },
  },
  "errors": undefined,
  "extensions": undefined,
  "http": Object {
    "headers": Headers {
      Symbol(map): Object {},
    },
  },
}
`;

exports[`graphql server queries park returns a park by its permalink 1`] = `
Object {
  "data": Object {
    "park": Object {
      "name": "Park One",
      "permalink": "park-one",
    },
  },
  "errors": undefined,
  "extensions": undefined,
  "http": Object {
    "headers": Headers {
      Symbol(map): Object {},
    },
  },
}
`;

exports[`graphql server queries parks returns a list of parks 1`] = `
Object {
  "data": Object {
    "parks": Array [
      Object {
        "name": "Park One",
        "permalink": "park-one",
      },
      Object {
        "name": "Park Two",
        "permalink": "park-two",
      },
    ],
  },
  "errors": undefined,
  "extensions": undefined,
  "http": Object {
    "headers": Headers {
      Symbol(map): Object {},
    },
  },
}
`;
